### PR TITLE
Add streaming onnx export for zipformer

### PR DESCRIPTION
**Pull request by yuekaizhang**
There are total 7 * (2+4+3+2+4) state tensors for onnx input, which is too much. I reduced the number of state tensors to four: len_cache, avg_cache, attn_cache(include cached_key, cached_val1, cached_val2), cnn_cache(include cached_conv1, cached_conv2) by adding a wrapper model on zipfomer.

### CODE REVIEW

**Comment by csukuangfj** [comment]
Could you also update 
https://github.com/k2-fsa/icefall/blob/master/.github/scripts/run-librispeech-pruned-transducer-stateless7-streaming-2022-12-29.sh
to add a test for onnx ?

You can use
https://github.com/k2-fsa/icefall/blob/8582b6e41acbd1258633492212c06589d1370960/.github/scripts/run-librispeech-pruned-transducer-stateless3-2022-05-13.sh#L30
as an example.

---

Also, please follow
https://github.com/k2-fsa/icefall/blob/8582b6e41acbd1258633492212c06589d1370960/.github/workflows/run-librispeech-pruned-transducer-stateless3-2022-05-13.yml#L42
to change
https://github.com/k2-fsa/icefall/blob/8582b6e41acbd1258633492212c06589d1370960/.github/workflows/run-librispeech-2022-12-29-stateless7-streaming.yml#L42

that is, add `github.event.label.name == 'onnx'`.

---

To activate the CI for onnx test, please add the label `onnx`
Screenshot 2023-01-11 at 15 59 42

**Comment by yuekaizhang** [comment]
> Hi @yuekaizhang , Seems there is no need to warp Decoder and Joiner for onnx export, just as [nonstreaming](https://github.com/k2-fsa/icefall/blob/master/egs/librispeech/ASR/pruned_transducer_stateless7/export.py#L285-L423). And, could you make the Encoder warper more flexible to support different `attention_dim`, `num_head`, even `cnn_module_kernels`, for example, someone may init zipformer as:
> 
> ```
> encoder = Zipformer(
>      num_features=80,
>      output_downsampling_factor=1,
>      encoder_dims=(384,512),
>      attention_dims=(256,256), # <---
>      encoder_unmasked_dims=(256,256),
>      zipformer_downsampling_factors=(4,8),
>      nhead=(4,8), # <---
>      feedforward_dim=(512,1024),
>      num_encoder_layers=(2,3),
>      dropout=0.1,
>      cnn_module_kernels=(31,15), # <---
>      pos_dim=4,
>      warmup_batches=4000.0,
> )
> ```
> 
> Thanks.
Hi,
Decoder and Joiner Wrapper are for triton, you may check the code comments. I think Encoder Wrapper just sets self.model=Zipformer. It has little to do with encoder hyperparameters.

**Comment by kobenaxie** [comment]
> > Hi @yuekaizhang , Seems there is no need to warp Decoder and Joiner for onnx export, just as [nonstreaming](https://github.com/k2-fsa/icefall/blob/master/egs/librispeech/ASR/pruned_transducer_stateless7/export.py#L285-L423). And, could you make the Encoder warper more flexible to support different `attention_dim`, `num_head`, even `cnn_module_kernels`, for example, someone may init zipformer as:
> > ```
> > encoder = Zipformer(
> >      num_features=80,
> >      output_downsampling_factor=1,
> >      encoder_dims=(384,512),
> >      attention_dims=(256,256), # <---
> >      encoder_unmasked_dims=(256,256),
> >      zipformer_downsampling_factors=(4,8),
> >      nhead=(4,8), # <---
> >      feedforward_dim=(512,1024),
> >      num_encoder_layers=(2,3),
> >      dropout=0.1,
> >      cnn_module_kernels=(31,15), # <---
> >      pos_dim=4,
> >      warmup_batches=4000.0,
> > )
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Thanks.
> 
> Hi, Decoder and Joiner Wrapper are for triton, you may check the code comments. I think Encoder Wrapper just sets self.model=Zipformer. It has little to do with encoder hyperparameters.

Thanks, if zipformer has different conv_kernel_size in different layers, the conv_cache can not cat together, as the last dimension  are not same, so conv cache also need pad ?

**Comment by yuekaizhang** [comment]
> > > Hi @yuekaizhang , Seems there is no need to warp Decoder and Joiner for onnx export, just as [nonstreaming](https://github.com/k2-fsa/icefall/blob/master/egs/librispeech/ASR/pruned_transducer_stateless7/export.py#L285-L423). And, could you make the Encoder warper more flexible to support different `attention_dim`, `num_head`, even `cnn_module_kernels`, for example, someone may init zipformer as:
> > > ```
> > > encoder = Zipformer(
> > >      num_features=80,
> > >      output_downsampling_factor=1,
> > >      encoder_dims=(384,512),
> > >      attention_dims=(256,256), # <---
> > >      encoder_unmasked_dims=(256,256),
> > >      zipformer_downsampling_factors=(4,8),
> > >      nhead=(4,8), # <---
> > >      feedforward_dim=(512,1024),
> > >      num_encoder_layers=(2,3),
> > >      dropout=0.1,
> > >      cnn_module_kernels=(31,15), # <---
> > >      pos_dim=4,
> > >      warmup_batches=4000.0,
> > > )
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Thanks.
> > 
> > 
> > Hi, Decoder and Joiner Wrapper are for triton, you may check the code comments. I think Encoder Wrapper just sets self.model=Zipformer. It has little to do with encoder hyperparameters.
> 
> Thanks, if zipformer has different conv_kernel_size in different layers, the conv_cache can not cat together, as the last dimension are not same, so conv cache also need pad ?

True. This PR is mainly for pretrained model 2022-12-29-stateless7-streaming. If "zipformer has different conv_kernel_size in different layers" this is the case, we could update onnx export in the future.

**Comment by csukuangfj** [comment]
Thanks for your contribution!
